File: src/server/sp/Web.js
"use strict";

var edge = require("edge");

var getWeb = edge.func(function(){/*
	#r "FakePoint.dll"
	
	using Microsoft.SharePoint;
	using System.Dynamic;

	async (input) => { 
        var site = SPContext.Current.Site;  			// TODO: Does not use correct site

        if(string.IsNullOrEmpty(input as string))
        {
			using(var web = site.OpenWeb())
	        {
		        return new {
			        ID = web.ID.ToString("B").ToUpper(), 	// TODO: move Formatting to Javascript
			        Url = web.Url
		        };
	        }
        }
        else
        {
        	var s = input as string;
            Guid id = new Guid();
            if (Guid.TryParse(s, out id))
            {
                using(var web = site.OpenWeb(id))
		        {
			        return new {
				        ID = web.ID.ToString("B").ToUpper(), 	// TODO: move Formatting to Javascript
				        Url = web.Url
			        };
		        }
            }
            else
            {
                using(var web = site.OpenWeb(s))
		        {
			        return new {
				        ID = web.ID.ToString("B").ToUpper(), 	// TODO: move Formatting to Javascript
				        Url = web.Url
			        };
		        }
            }
        }
    }
*/});

var Web = function (Url, callback) { // TODO: Should the contructor be usable? Would need tests
	var self = this;
	getWeb(Url, function getWebComplete(error, result){
		if(error) throw error;
        this.ID = result.ID;
		this.Url = result.Url; 
		this.Lists = {};
		callback(this);
	});
};

module.exports = Web;

File: nodepoint.Adpater/Web.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.SharePoint;
using System.Dynamic;

namespace nodepoint.Adpater
{
    class Web
    {
        public async Task<object> GetWeb(object input)
        {
            var site = SPContext.Current.Site;  			// TODO: Does not use correct site
            SPWeb web = await Task.Run(() =>
            {
                SPWeb result;
                if (string.IsNullOrEmpty(input as string))
                {
                    result = site.OpenWeb();
                }
                else
                {
                    var s = input as string;
                    Guid id = new Guid();
                    if (Guid.TryParse(s, out id))
                        result = site.OpenWeb(id);
                    else
                        result = site.OpenWeb(s);
                }

                return result;
            });
            
            return new
            {
                ID = web.ID.ToString("B").ToUpper(), 	// TODO: move Formatting to Javascript
                Title = web.Title,
                Url = web.Url,
                AllowUnsafeUpdates = true,
                Lists = await (new ListCollection()).GetLists(web)
            };
        }

    }
}
